eservice/lib/libpdo_enclave/contract_enclave.cpp

```diff
@@ -257,20 +257,25 @@ pdo_err_t ecall_HandleContractRequest(const uint8_t* inSealedSignupData,
             request.input_state_hash_,
             request.contract_id_hash_);
 
-        if(request.is_initialize())
+        if(request.is_initialize()) {
+            // validate contract compiler before saving to state for
```

Comment:
Let me start saying that this is just a clarification. I think the code works (and fails) as expected.

**Short version.**
Compiler and report checks are now before writing and after reading the state. Overall, 2 pairs of checks, with each pair separated in 2 files.
By focusing on guarding the contract execution (line 280), the code could equivalently implement 1 pair of checks in 1 file.

**Long version.**
Regarding "validate compiler ... before saving to state...", the comment suggests that the point is to write something good in the state, but it is not the case. In fact, the validation could be done afterwards. If it fails, there will be no state update, so nothing to commit.

Regarding the `ValidateContractCompiler` call/check here, (similar to a previous comment of mine) this is based on the eservice's untrusted data (i.e., the trusted compiler keys). So, such data determines whether something is written or not in the contract state. Although the effects are correct for an external observer, it should not be up to the eservice to determine that.

It is instead up the eservice to determine whether the execution should happen or not. The checks work fine because they are before line 280. That's where a single pair of compiler/report checks would be enough. Also, if successful, then whatever was written to, or read from, state was good; otherwise, no update, no commit. Besides consolidation, I think the two checks closer together would be clearer.
```
if(not accept all code)
    check compiler
    check report
fi
process request
```